### PR TITLE
Getting started with decorators in python

### LINKED ISSUE #2718: Getting Started with python decorators
### Proposed title of article
Getting Started with python decorators

### Introduction paragraph (2-3 paragraphs):
Pythons decorators are a design pattern that allows one to modify the functionality of an existing function  without change in the object structure. They are callable objects that can be used to  modify a function, method or a class. The decorators occur in the line just immediately before the object definition.

In this article we will look into python decorators and how to use them in python.

### Key takeaways: 
1. When to use decorators
2. How to create python decorators
3. Class decorators in python

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by mia-roberts**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
This fixes #2718

### CODE REVIEW

**Comment by mercymeave** [comment]
@mia-roberts  I will be reviewing this PR for my second test-review.

Can someone add me the labels please?

@ahmadmardeni1 , @hectorkambow

**Comment by mia-roberts** [comment]
This fixes #2718

**Comment by mercymeave** [comment]
Thank you Erick :wink:

File: content/articles/python-decorators/index.md [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+  - url: /engineering-education/python-decorators/hero.jpg
+    alt: Python decorators image
+---
+Python decorators are a design pattern that allows the developer to modify the structure of a function without changing the function's code structure. 
+<!--more-->
+This concept is useful when you want to change the behavior of a function, so rather than going in and changing the entire cod, you create a python decorator which will allow you to use a single line of code to change the behavior of all different functions.
+
+### Prerequisites
+Appreciating that everything in Python programming comes as an object is a key to understanding this concept. The reader should be well conversant with Python, although the explanations are simple and precise for beginners.
```
Comment (mercymeave) [inline content/articles/python-decorators/index.md:22]:
```suggestion
Appreciating that everything in Python programming comes as an object is a key to understanding this concept. The reader should be well conversant with Python, although the explanations are simple and precise for beginners.
```

Comment (mercymeave) [inline content/articles/python-decorators/index.md:22]: Make this point to come out clear  to the reader.

File: content/articles/python-decorators/index.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+Besides that, you need:
+- A working installation of Python on your computer
```
Comment (mercymeave) [inline content/articles/python-decorators/index.md:25]: Add python download link.

File: content/articles/python-decorators/index.md [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+- A code editor to try out the snippets
+
+### Understanding what decorators
+The concept of Python decorators is unclear to many people, even though it's simple. For this reason, you can consider Python decorators as high-level functions in Python which that takes another function as its argument and returns another function as a return value.
```
Comment (mercymeave) [inline content/articles/python-decorators/index.md:29]: Python dcorators being unclear is not a reason to make it a high level function.  Give a good reason as to why moajority of programmers do not understand python decorators.

File: content/articles/python-decorators/index.md [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+```py
+def another_function():
+    print('I am a computer science student')
+
+x = umbrella_function(another_function)
+```
+
+### Passing parameters to decorators
+Sometimes we will need to define a decorator function that accepts parameters. We can achieve this by passing the parameters to the wrapper function, which we then pass to the function that is being decorated.
```
Comment (mercymeave) [inline content/articles/python-decorators/index.md:170]:
```suggestion
Sometimes we may need to define a decorator function that accepts parameters. We can achieve this by passing the parameters to the wrapper function, which we then pass to the function that is being decorated.
```

File: content/articles/python-decorators/index.md [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+```bash
+another_function
+I am another function 
+umbrella_function
+I am the umbrella function
+```
+
+### Using multiple decorators
+Python allows the use of multiple decorators together. This process involves decorating a single function more than once with the same or different decorators. To use several decorators, we place the decorators just above the desired function. Python automatically does the chaining of the decorators.
```
Comment (mercymeave) [inline content/articles/python-decorators/index.md:257]:
```suggestion
Python allows the use of multiple decorators together. This process involves decorating a single function more than once with the same or different decorators.

  To use several decorators, we place the decorators just above the desired function. Python automatically does the chaining of the decorators. 
```

**Review by mercymeave — CHANGES_REQUESTED** [review]
Hello @mia-roberts 

Thank you for submitting your article for review. Upon conducting a prelimiary review, I noticed some minor mistakes and sugggeted corrections on the same.

Secondly, ensure that the hero image for the article is 600*400 pixels.


Kindly go through the suggestion and make appropriate changes then we move foraward.

**Comment by mia-roberts** [comment]
@mercymeave I have made the changes, kindly take a look.

**Review by mercymeave — APPROVED** [review]
Well done, I will proceed to tech review.

One of the content moderators will remove the revisions label.

@ahmadmardeni1 , @Ericgacoki

**Comment by mercymeave** [comment]
Thanks Erick

**Comment by mercymeave** [comment]
@mia-roberts , your article looks amazing with easy to follow snippets. One more thing before we finalize, add a table of content for easy navigatrion of your article.

**Comment by mia-roberts** [comment]
@mercymeave, updated, kindly check it out.

**Comment by mercymeave** [comment]
On it.

File: content/articles/python-decorators/index.md [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+Python decorators are a design pattern that allows the developer to modify the structure of a function without changing the function's code structure. 
+<!--more-->
+This concept is useful when you want to change the behavior of a function, so rather than going in and changing the entire cod, you create a python decorator which will allow you to use a single line of code to change the behavior of all different functions.
+
+### Table of content
+- [Prerequisites](#prerequisites)
+- [Understanding what decorators](#understanding-what-decorators)
+- [Core pillars of decorators](#core-pillars-of-decorators)
+- [Creating Python decorators](#creating-python-decorators)
+- [Passing parameters to decorators](#passing-parameters-to-ecorators)
```
Comment (mercymeave) [inline content/articles/python-decorators/index.md:26]:
```suggestion
- [Passing parameters to decorators](#passing-parameters-to-decorators)
```

File: content/articles/python-decorators/index.md [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+**Output:**
+
+```bash
+Arguments passed are: Mia, 18
+So, Mia is 18 years old, and she is a programmer
+```
+
+### Debugging Python decorators
+As observed, decorators wrap functions. The original function and its parameter are all masked by the wrapper. This issue poses a challenge on debugging and calls for the need to solve this challenge. Python provides a `functools.wraps()` decorator that copies the metadata from the undecorated function to the decorated wrapper function. Let's see how to do this.
```
Comment (mercymeave) [inline content/articles/python-decorators/index.md:207]:
```suggestion
As observed, decorators wrap functions. The original function and its parameter are all masked by the wrapper. This issue poses a challenge on debugging and calls for the need to solve this challenge. 

However, Python provides a `functools.wraps()` decorator that copies the metadata from the undecorated function to the decorated wrapper function. Let's see how to do this.
```

**Review by mercymeave — CHANGES_REQUESTED** [review]
@mia-roberts, Make the minor corrections then the article will be ready for final review.

**Comment by hectorkambow** [comment]
plagiarism check done

**Review by mercymeave — APPROVED** [review]

**Comment by mercymeave** [comment]
@mia-roberts  An amazing article :clap: 

In your coming articles, be sure to go over the minor suggestios and edits to polish them even further. Your articel is ready for final review.

@hectorkambow , @ahmadmardeni1, Would you please add me the PRC line and concuct a final review on this PR.

Thanks

**Comment by mercymeave** [comment]
@ahmadmardeni1 , You missed my name when you added the PRL. :smile:  Its Mercy Meave

**Comment by ahmadmardeni1** [comment]
Oh I'm so sorry! I copied the line and forgot to edit the first part. @mercymeave

**Comment by hectorkambow** [comment]
> Oh I'm so sorry! I copied the line and forgot to edit the first part. @mercymeave

Happens from time to time 👍  

😄

**Comment by mia-roberts** [comment]
@mia-roberts I think i must have missed it when making edits.

It is fine now.

**Comment by hectorkambow** [comment]
Overall great work @mia-roberts - very easy to follow along 🚀 

cc @mercymeave

**Pull request merged**